docs/design/sql/20-sql-partition-pruning.md

```diff
@@ -0,0 +1,309 @@
+| Related Jira                   | [HZ-1605](https://hazelcast.atlassian.net/browse/HZ-1605) |
+| Related Github issues          | _GH issue list_                                           |
+| Document Status / Completeness | IN PROGRESS                                               |
+| Requirement owner              | Sandeep Akhouri                                           |
+| Developer(s)                   | Krzystof Jamróz                                           |
+| Quality Engineer               | Isaac Sumner                                              |
+| Support Engineer               | _Support Engineer_                                        |
+| Technical Reviewers            | _Technical Reviewers_                                     |
+| Simulator or Soak Test PR(s)   | _Link to Simulator or Soak test PR_                       |
+
+### Background
+#### Description
+
+In practice many queries for large data sets do not need to scan all data.
+Very often data is partitioned and only limited subset of partitions needs to accessed during the query.
+Also some other operations like joins or aggregations can benefit from the knowledge about data partitioning.
+Using such knowledge makes it possible to run a query faster and using less resources (network, IO).
+This will make SQL queries eligible for partition pruning comparable in performance to 
+Predicate API queries using `PartitionPredicate`.
+
+Knowing which partitions are needed allows also to eliminate some members completely from the query execution.
+Such members do not have any data related to the query.
+
+
+
+
+#### Terminology
+
+| Term             | Definition                                       |
+|------------------|--------------------------------------------------|
+| item, entry, row | Single element of data which is treated as whole |
+|                  |                                                  |
+
+
+*Definition: Partitioning column*
+
+Partitioning column is a column/attribute that impacts partition to which given item is assigned.
+
+*Definition: Partitioning key*
+
+Partitioning key is a minimum (in terms of inclusion) set of partitioning columns that are sufficient
+to determine to which partition given item belongs.
+
+*Definition: Set of partitioning keys*
+
+Set of partitioning keys is defined as a set containing:
+- partitioning key defined by means of `AttributePartitioningStrategy` (if any)
+- `__key`
+
+_Clarification:_
+For IMap, partitioning columns may be attributes of IMap key. 
+If IMap uses `AttributePartitioningStrategy`, partitioning columns are functionally dependent on entire `__key`. 
+However for the sake of simplicity in the SQL optimizer we currently do not track functional dependencies
+and assume that both `__key` and `AttributePartitioningStrategy` define a partitioning key.
+Partition id calculation will take that into account and use a correct strategy for given IMap.
+
+*Definition: Partition-complete expression*
```

Comment:
@ivanthescientist you named this "Full partition filter", I do not like both names (mine is also not perfect), maybe there is some standard name for that?